// rest/controllers/posts.js
const db = require('db')
const errors = require('errors')

module.exports = {
    getPaginatedPosts: (req, res) => {
        res.sendWithPagination('Post', 5)
    },
    getLatestPost: (req, res) => {
        db.Post.findOne({ order:'`created_at` DESC'})
        .then((post) => {
            res.json(post)
        })
        .catch(errors.handleRestError(req, res))
    },
    getPostById: (req, res) => {
        db.Post.findById(req.params.id)
        .then((post) => {
            res.json(post)
        })
        .catch(errors.handleRestError(req, res))
    },
    createPost: (req, res) => {
        db.Post.create({
            title: req.body.title,
            content: req.body.content
        })
        .then((post) => {
            res.json(post)
        })
        .catch(errors.handleRestError(req, res))
    },
    editPost: (req, res) => {
        const fieldsToUpdate = {
            content: req.body.content,
            title: req.body.title
        }

        db.Post.findById(req.params.id)
        .then((post) => {
            return post.update(fieldsToUpdate)
        })
        .then((updatedPost) => {
            res.json(updatedPost)
        })
        .catch(errors.handleRestError(req, res))
    }
}
