// rest/controllers/projects.js
const db = require('db')
const errors = require('errors')

module.exports = {
    getPaginatedProjects: (req, res) => {
        res.sendWithPagination('Project', 5)
    },
    getProjectById: (req, res) => {
        db.Project.findById(req.params.id)
        .then((project) => {
            res.json(project)
        })
        .catch(errors.handleRestError(req, res))
    },
    createProject: (req, res) => {
        db.Project.create({
            title: req.body.title,
            content: req.body.content
        })
        .then((project) => {
            res.json(project)
        })
        .catch(errors.handleRestError(req, res))
    },
    editProject: (req, res) => {
        const fieldsToUpdate = {
            content: req.body.content,
            title: req.body.title
        }

        db.Project.findById(req.params.id)
        .then((project) => {
            return project.update(fieldsToUpdate)
        })
        .then((updatedProject) => {
            res.json(updatedProject)
        })
        .catch(errors.handleRestError(req, res))
    }
}
